FabNESO/tasks.py

```diff
@@ -252,3 +267,209 @@ def neso_ensemble(
+        # without it creating a new xwindow?
+        "plot": False,
+        # Run with a log file
+        "log_file_name": logfile_path,
+    }
+
+    # Make an instance of VBMC
+    vbmc = pyvbmc.VBMC(
+        lambda theta: log_density(theta, config_dict),
+        theta_0,
+        lower_bounds,
+        upper_bounds,
+        plausible_lower_bounds,
+        plausible_upper_bounds,
+        options=options,
+    )
+
+    # Run the vbmc instance
+    vp, results = vbmc.optimize()
+
+    # Print the final output of vbmc to the logfile
+    with logfile_path.open("a+") as logfile:
+        logfile.write(pyvbmc.formatting.format_dict(results))
+
+    # Save the final output posterior
+    vbmc.vp.save(output_directory / f"final_posterior_{timestamp}.pkl")
+
+
+def run_instance_return_field(
+    config_dict: dict[str, Any],
+) -> dict[str, np.ndarray]:
+    """Run a single instance of the NESO solver and return the observed_field."""
+    # If we're running remotely, tell the submission to wait until done
+    if "archer2" in fab.env.remote:
+        fab.update_environment(
+            {
+                "job_dispatch": "cd /work/$project/$project/$username ; sbatch --wait",
+            }
+        )
+    neso(
+        config=config_dict["config"],
+        solver=config_dict["neso_args"]["neso_solver"],
+        conditions_file_name=config_dict["neso_args"]["neso_conditions_file"],
+        mesh_file_name=config_dict["neso_args"]["neso_mesh_file"],
+        processes=config_dict["neso_args"]["cores"],
+        nodes=config_dict["neso_args"]["nodes"],
+        cpus_per_process=config_dict["neso_args"]["cpuspertask"],
+        wall_time=config_dict["neso_args"]["job_wall_time"],
+        create_missing_parameters=True,
+        **config_dict["para_overrides"],
+    )
+    fab.fetch_results()
+    local_results_dir = Path(fab.env.job_results_local) / template(
+        fab.env.job_name_template
+    )
+    final_line_field_step = (
+        int(
+            list_parameter_values(
+                Path(fab.find_config_file_path(config_dict["config"]))
+                / str(config_dict["neso_args"]["neso_conditions_file"]),
+                "particle_num_time_steps",
+            )[0]
+        )
+        - config_dict["para_overrides"]["line_field_deriv_evaluations_step"]
+    )
+    return read_hdf5_datasets(
+        local_results_dir / "Electrostatic2D3V_line_field_deriv_evaluations.h5part",
+        {
+            "x": f"Step#{final_line_field_step}/x",
+            "field_value": f"Step#{final_line_field_step}/FIELD_EVALUATION_0",
+        },
+    )
+
+
+def log_density(
+    theta: list[float],
+    config_dict: dict[str, Any],
+) -> list:
+    """Run an instance of the neso task and return the log_joint_density."""
+    parameters = dict(
+        zip(config_dict["parameters_to_scan"].keys(), theta, strict=True),
+        **config_dict["para_overrides"],
+    )
+
+    config_dict["para_overrides"] = parameters
```

Comment:
Overwriting the previous value in `config_dict` could be a bit error prone as I think there will be a single `config_dict` instance which is shared across all calls to this function. The unpacking of `**config_dict["para_overrides"]` in the definition of `parameters` I think means any values written to `para_overrides` in a previous iteration might end up being used in place of the values extracted from `theta` (as `dict([("a", 1)], **{"a": 2})` evaluates to `{"a": 2}`). I would instead pass the `parameters` dict as a separate argument to `run_instance_return_field`  and use in place of the ` **config_dict["para_overrides"]` expression in the call to `neso`.

Reply:
So I did consider this, and from what I could tell it was behaving correctly, but I think you're right in that we should treat it in a safer way. To that point I've added in the parameter overrides as an additional argument in `run_instance_return_field` as suggested.